File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        freq = {}
        ans = 0
        left = 0
        for right in range(len(nums)):
            if nums[right] not in freq:
                freq[nums[right]] = 1
            else:
                freq[nums[right]] += 1
            
            while left < right and freq[nums[right]] > k:
                freq[nums[left]] -= 1
                left += 1
            
            ans = max(ans, right - left + 1)
        
        return ans

# array, hash map, sliding window
# time: O(n)
# space: O(n)

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.go
func maxSubarrayLength(nums []int, k int) int {
    m := map[int]int{}
    ans := 0
    left := 0
    for right := 0; right < len(nums); right++ {
        m[nums[right]]++

        if m[nums[right]] > k {
            ans = max(ans, right - left)

            for left < right && m[nums[right]] > k {
                m[nums[left]]--
                left++
            }
        }

        if right == len(nums) - 1 {
            ans = max(ans, right - left + 1)
        }
    }

    return ans
}

// array, hash map, sliding window
// time: O(n)
// space: O(n)